libsql-wal/src/io/file.rs

```diff
@@ -0,0 +1,246 @@
+use std::fs::File;
+use std::future::Future;
+use std::io::{self, ErrorKind, IoSlice, Result, Write};
+
+use super::buf::{IoBuf, IoBufMut};
+
+pub trait FileExt: Send + Sync + 'static {
+    fn len(&self) -> io::Result<u64>;
+    fn write_all_at(&self, buf: &[u8], offset: u64) -> Result<()>;
+    fn write_at_vectored(&self, bufs: &[IoSlice], offset: u64) -> Result<usize>;
+    fn write_at(&self, buf: &[u8], offset: u64) -> Result<usize>;
+
+    fn read_exact_at(&self, buf: &mut [u8], offset: u64) -> Result<()>;
+
+    fn sync_all(&self) -> Result<()>;
+
+    fn set_len(&self, len: u64) -> Result<()>;
+
+    fn cursor(&self, offset: u64) -> Cursor<Self>
+    where
+        Self: Sized,
+    {
+        Cursor {
+            file: self,
+            offset,
+            count: 0,
+        }
+    }
+
+    fn read_exact_at_async<B: IoBufMut + Send + 'static>(
+        &self,
+        buf: B,
+        offset: u64,
+    ) -> impl Future<Output = (B, Result<()>)> + Send;
+    fn write_all_at_async<B: IoBuf + Send + 'static>(
+        &self,
+        buf: B,
+        offset: u64,
+    ) -> impl Future<Output = (B, Result<()>)> + Send;
+}
+
+impl FileExt for File {
+    fn write_all_at(&self, buf: &[u8], offset: u64) -> Result<()> {
+        let mut written = 0;
+
+        while written != buf.len() {
+            written += nix::sys::uio::pwrite(self, &buf[written..], offset as _)?;
```

Comment:
dont you need to write at offset + written?

Reply:
hmm yeah, I need to write a test that catches that

Reply:
I was unable to write a test to verify that, no matter how big the payload, it's writing everything in one go...